client/rpmtrans.c

```diff
@@ -577,24 +596,32 @@ doCheck(PTDNFRPMTS pTS)
                     if (rpmProblemGetType(prob) == RPMPROB_REQUIRES)
                     {
                         uint32_t dwError = 0;
+
                         dwError = TDNFAllocateString(rpmProblemGetStr(prob), &pErrorStr);
                         BAIL_ON_TDNF_ERROR(dwError);
 
                         dwError = TDNFDetectPreTransFailure(pTS->pTS, pErrorStr);
                         BAIL_ON_TDNF_ERROR(dwError);
+
+                        TDNF_SAFE_FREE_MEMORY(pErrorStr);
                     }
                 }
-                rpmProblemFree(prob);
             }
-            rpmpsFreeIterator(psi);
-            nResult = ERROR_TDNF_RPM_CHECK;
         }
     }
 cleanup:
-    TDNF_SAFE_FREE_MEMORY(pErrorStr);
-    return nResult;
+    return;
+
 error:
-    nResult = ERROR_TDNF_RPM_CHECK;
+    TDNF_SAFE_FREE_MEMORY(pErrorStr);
+    if (psi)
+    {
+        rpmpsFreeIterator(psi);
+    }
+    if (ps)
+    {
+        rpmpsFree(ps);
+    }
```

Comment:
I think we should always free these, not just in the error case.

Reply:
You're absolutely correct. My apologies, that was definitely unintended. I missed that with the churn from the last round of changes--thanks for catching it.

I've pushed an update addressing this.